pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/JsonExtractIndexTransformFunction.java

```diff
@@ -0,0 +1,256 @@
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.core.operator.transform.function;
+
+import java.math.BigDecimal;
+import java.util.List;
+import java.util.Map;
+import org.apache.pinot.core.operator.ColumnContext;
+import org.apache.pinot.core.operator.blocks.ValueBlock;
+import org.apache.pinot.core.operator.transform.TransformResultMetadata;
+import org.apache.pinot.segment.spi.index.reader.JsonIndexReader;
+import org.apache.pinot.segment.spi.index.reader.JsonIndexReaderContext;
+import org.apache.pinot.spi.data.FieldSpec.DataType;
+
+
+/**
+ * The <code>JsonExtractIndexTransformFunction</code> provides the same behavior as JsonExtractScalar, with the
+ * implementation changed to read values from the JSON index. For large JSON blobs this can be faster than parsing
+ * GBs of JSON at query time. For small JSON blobs/highly filtered input this is generally slower than the *scalar
+ * implementation. The inflection point is highly dependent on the number of docs remaining post filter.
+ */
+public class JsonExtractIndexTransformFunction extends BaseTransformFunction {
+  public static final String FUNCTION_NAME = "jsonExtractIndex";
+
+  private TransformFunction _jsonFieldTransformFunction;
+  private String _jsonPathString;
+  private TransformResultMetadata _resultMetadata;
+  private JsonIndexReader _jsonIndexReader;
+  private Object _defaultValue;
+  private JsonIndexReaderContext _jsonIndexReaderContext;
+
+  @Override
+  public String getName() {
+    return FUNCTION_NAME;
+  }
+
+  @Override
+  public void init(List<TransformFunction> arguments, Map<String, ColumnContext> columnContextMap) {
+    // Check that there are exactly 3 or 4 arguments
+    if (arguments.size() < 3 || arguments.size() > 4) {
+      throw new IllegalArgumentException(
+          "Expected 3/4 arguments for transform function: jsonExtractIndex(jsonFieldName, 'jsonPath', 'resultsType',"
+              + " ['defaultValue'])");
+    }
+
+    TransformFunction firstArgument = arguments.get(0);
+    if (firstArgument instanceof LiteralTransformFunction || !firstArgument.getResultMetadata().isSingleValue()) {
+      throw new IllegalArgumentException(
+          "The first argument of jsonExtractIndex transform function must be a single-valued column or a transform "
+              + "function");
+    }
+    if (firstArgument instanceof IdentifierTransformFunction) {
+      String columnName = ((IdentifierTransformFunction) firstArgument).getColumnName();
+      if (columnContextMap.containsKey(columnName)) {
```

Comment:
(minor) This check is redundant. Identifier is guaranteed to be in the `columnContextMap`